README
#!/usr/bin/env perl
#
# Usage: perl README | w3m -T text/html
#

use lib 'lib';
use Markapl;

sub readme {
    html {
        head {
            title { "Markapl - Markup as Perl" };
        };
        body {
            h2 { "DESCRIPTION" };

            p("#description") {
                "Markapl is the way to generate HTML with (augmented) Perl syntax. It's inspired by both Template::Declare and Markaby."
            };

            h2 { "INSTALLATION" };

            p { "To install this module, run the following commands:" };

            pre {
                q{
    perl Makefile.PL
    make
    make test
    make install}
             };

            h2 { "COPYRIGHT AND LICENCE" };

            p { "Copyright (C) 2008, Kang-min Liu" };

            p {
                outs "This is free software, licensed under: ";
                strong { "The MIT (X11) License" };
            };
        }
    }
}

print main->render("readme");



Changes
# -*- yaml -*-
# Revision history for Markapl

0.13:
  - Allow extra [-:\.] case in id & class according to SGML basic types. by gslin++

0.12:
  - Minor mantainence updates.

0.11:
  - Several elements now produce outputs without closing tags as they are supposed to be.
  
0.10:
  - Small refactoring

0.09:
  - Increase the compatibility with Jifty::Web::PageRegion

0.08:
  - A maintainance release for the compatibility with JiftyX::Markapl
  - documentation fixes from dbrunton++ and rafl++
  - various internal fixes from douglashunter++

0.07:
  - Tag blocks do not need to have semi-colon at the end from now. 

0.06:
  - Improved the code to pass attribute with interpolate variables.
  - Less dependencies

0.05:
  - allow passing variables to render.
  - implement package-specific template stash

0.04:
  - even more code refactoring.
  - list html4 tags in Markapl::Tags
  - Add several very useful helpers

0.03:
  - element attribute values can now contains variables.

0.02:
  - mostly just code refactoing
  - documentation 
  - add 'template' funtion that defines template sub but can be named like "/foo.html" as in Template::Declare.

0.01:
  - Initial release.



t/lib/MyTestView.pm
package MyTestView;

use Markapl;

template "/foo.html" => sub {
    outs "foo.html output";
};

template "/foo/bar.html" => sub {
    outs "bar.html output";
};


template '/rowcell.html' => sub {
    table {
        row {
            cell { "HI" }
        }
    }
};

sub linklist {
    my @link = @_;
    ul {
        for my $l (@link) {
            li {
                a( href => $l->[1] ) { $l->[0] }
            };
        }
    };
}

template '/linklist.html' => sub {
    html {
        head { title { "Test" } };
        body {
            h1 { "Hi" };

            linklist(
                ["Currency", "/currency"],
                ["Foo", "/foo"],
            );
        }
    };
};

template '/var-in-attribute.html' => sub {
    my $root = "/foo";
    ul {
        li { a(href => $root) { "Home" }; };
        li { a(href => "$root/fun") { "Have Fun" }; };
        li { a(href => "$root/logout") { "Logout" }; };
    }
};

template '/perlbuiltin.html' => sub {
    html_q { "Here is a short quotation here is a short quotation" };
};

template '/stash1.html' => sub {
    my $title = get("title");
    h1 { $title };
};

1;



t/lib/MyTestView2.pm
package MyTestView2;
use strict;
use warnings;

use Markapl;
use Markapl::Helpers;

template '/helper-heads-cells' => sub {
    table {
	row { heads qw(Name Age Weight Height) };
	row { cells qw(Gugod 29 100kg 170cm)   };
    }
};

template '/helper-THs-TDs' => sub {
    table {
	row { THs qw(Name Age Weight Height) };
	row { TDs qw(Gugod 29 100kg 170cm)   };
    }
};

template '/helper-ths-tds' => sub {
    table {
	row { ths qw(Name Age Weight Height) };
	row { tds qw(Gugod 29 100kg 170cm)   };
    }
};

1;


t/lib/MyTestView3.pm
package MyTestView3;

use Markapl;

template "/var1.html" => sub {
    my ($self, %vars) = @_;
    h1(".title") { outs $vars{title} };
};

template '/stash1.html' => sub {
    my $title = get("title");
    h1 { $title };
};

1;


lib/Markapl/Tags.pm
package Markapl::Tags;

use strict;
use warnings;

our $VERSION = "0.01";

sub w3c_html4 {
    qw(
       a abbr acronym address applet area b base basefont bdo big
       blockquote body br button caption center cite code col colgroup
       dd del dfn dir div dl dt em fieldset font form frame frameset
       h1 h2 h3 h4 h5 h6 head hr html i iframe img input ins isindex
       kbd label legend li link map menu meta noframes noscript object
       ol optgroup option p param pre q s samp script select small
       span strike strong style sub sup table tbody td textarea tfoot
       th thead title tr tt u ul var
    );
}

sub html4_perl_common {
    qw(link map q s sub tr);
}

sub html {
    # removed common ones. add replacements
    qw(
       a abbr acronym address applet area b base basefont bdo big
       blockquote body br button caption center cite code col colgroup
       dd del dfn dir div dl dt em fieldset font form frame frameset
       h1 h2 h3 h4 h5 h6 head hr html i iframe img input ins isindex
       kbd label legend li menu meta noframes noscript object
       ol optgroup option p param pre samp script select small
       span strike strong style sup table tbody td textarea tfoot
       th thead title tt u ul var

       base html_link html_map html_q html_s html_sub html_tr
       cell row
    );
}

sub html_inline {
    qw( hr img input meta );
}

1;

1;

=head1 NAME

Markapl::Tags - List of tags.

=head1 DESCRIPTION

This module is used internally to get a list of HTML tags.  However,
Markapl users may call C<Markapl::Tags->html> to retrieve the list of
supported html tags. (Or even modify it)

Invoke C<perldoc -m Markapl::Tags> too see the full list of tag names.

=head1 FUNCTIONS

=over 4

=item html

Return the list of supported html tags. The following tags are in the
returned list but they are not standard html tags:

  html_link html_map html_q html_s html_sub html_tr cell row

Because their corresponding html tag names conflicts with perl builtin
functions.

=item w3c_html4

Return the list of standard html tag names. This is just here for
references.

=back

=head1 AUTHOR

Kang-min Liu  C<< <gugod@gugod.org> >>

=head1 LICENCE AND COPYRIGHT

Copyright (c) 2008, Kang-min Liu C<< <gugod@gugod.org> >>.

This is free software, licensed under:

    The MIT (X11) License

=head1 DISCLAIMER OF WARRANTY

BECAUSE THIS SOFTWARE IS LICENSED FREE OF CHARGE, THERE IS NO WARRANTY
FOR THE SOFTWARE, TO THE EXTENT PERMITTED BY APPLICABLE LAW. EXCEPT WHEN
OTHERWISE STATED IN WRITING THE COPYRIGHT HOLDERS AND/OR OTHER PARTIES
PROVIDE THE SOFTWARE "AS IS" WITHOUT WARRANTY OF ANY KIND, EITHER
EXPRESSED OR IMPLIED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE. THE
ENTIRE RISK AS TO THE QUALITY AND PERFORMANCE OF THE SOFTWARE IS WITH
YOU. SHOULD THE SOFTWARE PROVE DEFECTIVE, YOU ASSUME THE COST OF ALL
NECESSARY SERVICING, REPAIR, OR CORRECTION.

IN NO EVENT UNLESS REQUIRED BY APPLICABLE LAW OR AGREED TO IN WRITING
WILL ANY COPYRIGHT HOLDER, OR ANY OTHER PARTY WHO MAY MODIFY AND/OR
REDISTRIBUTE THE SOFTWARE AS PERMITTED BY THE ABOVE LICENCE, BE
LIABLE TO YOU FOR DAMAGES, INCLUDING ANY GENERAL, SPECIAL, INCIDENTAL,
OR CONSEQUENTIAL DAMAGES ARISING OUT OF THE USE OR INABILITY TO USE
THE SOFTWARE (INCLUDING BUT NOT LIMITED TO LOSS OF DATA OR DATA BEING
RENDERED INACCURATE OR LOSSES SUSTAINED BY YOU OR THIRD PARTIES OR A
FAILURE OF THE SOFTWARE TO OPERATE WITH ANY OTHER SOFTWARE), EVEN IF
SUCH HOLDER OR OTHER PARTY HAS BEEN ADVISED OF THE POSSIBILITY OF
SUCH DAMAGES.


lib/Markapl/Helpers.pm
package Markapl::Helpers;
use strict;
use Markapl;

use Exporter::Lite;
our $VERSION = "0.01";
our @EXPORT = qw(THs TDs ths tds heads cells linklist);

use Sub::Alias;

sub linklist {
    my @link = @_;
    ul {
        while(@link) {
            my $label = shift @link;
            my $url = shift @link;
            li {
                a(href => $url) { $label };
            }
        }
    }
}

sub heads {
    my @heads = @_;
    row { for (@heads) { th { $_ } } }
}

sub cells {
    my @cells = @_;
    row { for (@cells) { cell { $_ } } }
}

alias THs => 'heads';
alias ths => 'heads';
alias TDs => 'cells';
alias tds => 'cells';

1;

=head1 NAME

Markapl::Helpers - Practical helper methods

=head1 DESCRIPTION

This module exports several useful functions that are meant to be used
inside templates.

=head1 FUNCTIONS

=over 4

=item linklist($label1 => $url1, $label2 => $url2, ...)

Take pairs of labels and urls, return a structure like this:

    <ul>
      <li><a href="$url1">$label1</a></li>
      <li><a href="$url2">$label2</a></li>
      ...
    </ul>

=item cells($item1, $item2, ...)

Wrap C<< <td></td> >> tag around each item in the argument list.
Useful when showing tabular data. For example:

    table {
        while(my @data = $dbh->fetchrow_array) {
          cells(@data);
        }
    }

=item heads($item1, $item2, ...)

Similar to C<cells>, but output seris of C<< <th></th> >> tag instead.
For example:

    table {
        heads( qw(Name Age Gender Height Weight) );
        while(my @data = $dbh->fetchrow_array) {
          cells(@data);
        }
    }

=back

=head1 AUTHOR

Kang-min Liu  C<< <gugod@gugod.org> >>

=head1 LICENCE AND COPYRIGHT

Copyright (c) 2008, Kang-min Liu C<< <gugod@gugod.org> >>.

This is free software, licensed under:

    The MIT (X11) License

=head1 DISCLAIMER OF WARRANTY

BECAUSE THIS SOFTWARE IS LICENSED FREE OF CHARGE, THERE IS NO WARRANTY
FOR THE SOFTWARE, TO THE EXTENT PERMITTED BY APPLICABLE LAW. EXCEPT WHEN
OTHERWISE STATED IN WRITING THE COPYRIGHT HOLDERS AND/OR OTHER PARTIES
PROVIDE THE SOFTWARE "AS IS" WITHOUT WARRANTY OF ANY KIND, EITHER
EXPRESSED OR IMPLIED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE. THE
ENTIRE RISK AS TO THE QUALITY AND PERFORMANCE OF THE SOFTWARE IS WITH
YOU. SHOULD THE SOFTWARE PROVE DEFECTIVE, YOU ASSUME THE COST OF ALL
NECESSARY SERVICING, REPAIR, OR CORRECTION.

IN NO EVENT UNLESS REQUIRED BY APPLICABLE LAW OR AGREED TO IN WRITING
WILL ANY COPYRIGHT HOLDER, OR ANY OTHER PARTY WHO MAY MODIFY AND/OR
REDISTRIBUTE THE SOFTWARE AS PERMITTED BY THE ABOVE LICENCE, BE
LIABLE TO YOU FOR DAMAGES, INCLUDING ANY GENERAL, SPECIAL, INCIDENTAL,
OR CONSEQUENTIAL DAMAGES ARISING OUT OF THE USE OR INABILITY TO USE
THE SOFTWARE (INCLUDING BUT NOT LIMITED TO LOSS OF DATA OR DATA BEING
RENDERED INACCURATE OR LOSSES SUSTAINED BY YOU OR THIRD PARTIES OR A
FAILURE OF THE SOFTWARE TO OPERATE WITH ANY OTHER SOFTWARE), EVEN IF
SUCH HOLDER OR OTHER PARTY HAS BEEN ADVISED OF THE POSSIBILITY OF
SUCH DAMAGES.


lib/Markapl/TagHandlers.pm
package Markapl::TagHandlers;

use strict;
use warnings;
use Devel::Declare ();
use B::Hooks::EndOfScope;

our $VERSION = 0.03;

our ($Declarator, $Offset);

sub skip_declarator {
    $Offset += Devel::Declare::toke_move_past_token($Offset);
}

sub skipspace {
    $Offset += Devel::Declare::toke_skipspace($Offset);
}

sub strip_name {
    skipspace;
    if (my $len = Devel::Declare::toke_scan_word($Offset, 1)) {
        my $linestr = Devel::Declare::get_linestr();
        my $name = substr($linestr, $Offset, $len);
        substr($linestr, $Offset, $len) = '';
        Devel::Declare::set_linestr($linestr);
        return $name;
    }
    return;
}

sub strip_proto {
    skipspace;

    my $linestr = Devel::Declare::get_linestr();
    if (substr($linestr, $Offset, 1) eq '(') {
        my $length = Devel::Declare::toke_scan_str($Offset);
        my $proto = Devel::Declare::get_lex_stuff();
        Devel::Declare::clear_lex_stuff();
        $linestr = Devel::Declare::get_linestr();
        substr($linestr, $Offset, $length) = '';
        Devel::Declare::set_linestr($linestr);
        return $proto;
    }
    return;
}

sub shadow {
    my $pack = Devel::Declare::get_curstash_name;
    Devel::Declare::shadow_sub("${pack}::${Declarator}", $_[0]);
}

sub make_proto_unwrap {
    my ($proto) = @_;
    return (defined($proto) && length($proto)) ? "($proto);" : "";
}

sub inject_if_block {
    my $inject = shift;
    skipspace;
    my $linestr = Devel::Declare::get_linestr;
    if (substr($linestr, $Offset, 1) eq '{') {
        substr($linestr, $Offset+1, 0) = $inject;
        Devel::Declare::set_linestr($linestr);
    }
}

sub inject_before_block {
    my $inject = shift;
    skipspace;
    my $linestr = Devel::Declare::get_linestr;
    if (substr($linestr, $Offset, 1) eq '{') {
        substr($linestr, $Offset, 0) = $inject;
        Devel::Declare::set_linestr($linestr);
    }
}

sub inject_empty_block {
    my $inject = "{}";
    skipspace;
    my $linestr = Devel::Declare::get_linestr;
    if (substr($linestr, $Offset, 1) eq ';') {
        substr($linestr, $Offset, 0) = $inject;
        Devel::Declare::set_linestr($linestr);
    }
}

sub inject_scope {
    on_scope_end {
        my $linestr = Devel::Declare::get_linestr;
        my $offset = Devel::Declare::get_linestr_offset;
        substr($linestr, $offset, 0) = ';';
        Devel::Declare::set_linestr($linestr);
    };
}

my %alt = (
    'cell'      => 'td',
    'row'       => 'tr',
    'html_base' => 'base',
    'html_link' => 'link',
    'html_map'  => 'map',
    'html_q'    => 'q',
    'html_s'    => 's',
    'html_sub'  => 'sub',
    'html_tr'   => 'tr',
);

use String::BufferStack;

sub _tag {
    my ($tag, $attr, $block, $in_closure) = @_;

    my $original_buffer;

    if ($in_closure) {
        $original_buffer = Markapl->buffer;
        Markapl->buffer( String::BufferStack->new( out_method => sub { join("", @_) }) );
        Markapl->buffer->push;
    }

    if (grep { $tag eq $_ } Markapl::Tags->html_inline) {
        Markapl->buffer->append("<${tag}${attr}>");
    }
    else {
        Markapl->buffer->append("<${tag}${attr}>");
        Markapl->buffer->append(
            join '', map {
                ref($_) && $_->isa('Markapl::Tag') ? $_->() : ($_||"")
            } $block->()
        ) if defined $block && ref($block) eq 'CODE';
        Markapl->buffer->append("</$tag>");
    }

    if ($original_buffer) {
        my $output = Markapl->buffer->pop;
        Markapl->buffer($original_buffer);
        return $output;
    }

    return '';
}

sub tag_parser_for {
    my ($tag) = @_;
    $tag = $alt{$tag} if defined($alt{$tag});

    return sub {
        local ($Declarator, $Offset) = @_;

        my $offset_before = $Offset;
        skip_declarator;

        # This means that current declarator is in a hash key.
        # Don't shadow sub in this case
        return if $Offset == $offset_before;

        my $name = strip_name;
        my $proto = strip_proto;

        inject_if_block("BEGIN { Markapl::TagHandlers::inject_scope };");
        inject_empty_block;

        if (defined($proto)) {
            inject_before_block("$proto, sub");
        }
        else {
            inject_before_block("sub");
        }

        shadow(
            sub {
                my $block = pop;
                my @attr = @_;

                my $attr = "";

                if (@attr == 1) {
                    my $css = $attr[0];
                    while ($css =~ /([\#\.])([A-Za-z][-:\.\w]*)/g) {
                        if ($1 eq '#') {
                            $attr .= qq{ id="$2"};
                        } else {
                            $attr .= qq{ class="$2"};
                        }
                    }
                } else {
                    my ($k, $v) = (shift @attr, shift @attr);
                    while ($k && $v) {
                        $attr .= " $k=\"$v\"";
                        ($k, $v) = (shift @attr, shift @attr);
                    }
                }

                if (defined wantarray and not wantarray) {
                    my $sub = sub {
                        _tag($tag, $attr, $block, 1);
                    };
                    bless $sub, 'Markapl::Tag';
                    return $sub;
                }
                _tag($tag, $attr, $block);
                return '';
            }
        );

    }
}

1;


lib/Markapl.pm
package Markapl;
use strict;
use warnings;
use Devel::Declare ();
use Sub::Install qw(install_sub);
use Markapl::Tags;
use Markapl::TagHandlers;
use String::BufferStack;
use HTML::Entities;

use 5.008;
our $VERSION = "0.13";

my $buffer;
sub buffer {
    $buffer = $_[1] if $_[1];
    $buffer ||= String::BufferStack->new( out_method => sub { join("", @_) });
}

sub template {
    my ($name, $code) = @_;
    my $caller = caller;
    install_sub({
        code => $code,
        into => $caller,
        as => $name
    });
}

sub outs_raw($) {
    my $str = shift;
    Markapl->buffer->append( $str );
    return '';
}

sub outs($) {
    my $str = shift;
    outs_raw encode_entities($str, '<>&"');
}

sub render {
    my ($self, $template, @vars) = @_;

    if (my $sub = $self->can($template)) {
        Markapl->buffer->push;
        $sub->($self, @vars);
        my $buf = Markapl->buffer->pop;
        Markapl->buffer->flush_output;
        return $buf
    } else {
        require Carp;
        Carp::croak( "no such template: $template in $self" );
    }
}

sub set {
    my ($self, $name, $value) = @_;

    if ($self->can("stash")) {
        $self->stash->{$name} = $value;
    }
    elsif ((my $caller = caller)->can("stash")) {
        $caller->stash->{$name} = $value;
    }
}

sub get {
    my ($self, $name) = @_;
    if (defined($name) && $self->can("stash")) {
        return $self->stash->{$name} || ""
    }
    $name = $self, $self = caller;
    $self->get($name);
}

sub import {
    my ($class) = @_;
    my $caller = caller;

    for my $name (qw(render outs outs_raw template get set)) {
        install_sub({
            code => $name,
            into => $caller,
        });
    }

    my $config = {};
    my $code_str = qq{package $caller; my \$stash = {}; sub stash { \$stash } };

    for my $tag ( Markapl::Tags->html() ) {
        $code_str .= qq{sub $tag (&);};
        $config->{$tag} = {
            const => Markapl::TagHandlers::tag_parser_for($tag)
        }
    }

    eval $code_str;

    Devel::Declare->setup_for($caller, $config);
}

1;
__END__

=head1 NAME

Markapl - Markup as Perl

=head1 VERSION

This document describes Markapl version 0.08

=head1 SYNOPSIS

    package MyView;
    use Markapl;

    template '/a/page.html' => sub {
        h1("#title") { "Hi" };
        p(".first") { "In the begining, lorem ipsum...." };
        p(style => "color: red;") { "But...." };
    }

    package main;
    my $output = MyView->render("/a/page.html");

=head1 TUTORIAL

Here's a short guide how to use this module. You can skip this
tutorial section if you're already using L<Template::Declare>, since
it's exactly the same.

First of all, you need a sole package for defining your templates,
let's call it "MyView" in the example.  Then you C<use Markapl> (but
not use base), which installs many helper subroutines into your
"MyView" package for declaring and rendering templates:


    package MyView;
    use Markapl;

To define a template, use C<template> function like this:

    template '/page.html' => sub {
        h1("#title") { "Hi" };
        p(".first") { "In the begining, lorem ipsum...." };
        p(style => "color: red;") { "But...." };
    }

To render it, call C<render> function:

    my $out = MyView->render("/page.html");

Besides these two functions, C<Markapl> also exports about another 120
tag functions that are named after HTML tags, for exapmle, h1, h2, h3,
div, p, and span. Almost all HTML tags are defined as a function.

In your template, if you say:

    h1 { "Hi" };

It'll be rendered as:

    <h1>Hi</h1>

The block after C<h1> is an anonymous sub-routine, and the return
value of which will become the content of the C<h1> tag.

If you want to add attributes to the tag, do it like this:

    div(id => "example", class => "lipsum") { "Lorem ipsum" };

Alternatively, you can use CSS selector syntax to quickly defined
id and class attribute:

    div("#example") { "Lorem ipsum" };

That only works when the attribute list contain exactly one string inside.

A special function C<outs> need to be used to concatenate strings with
inline elements:

    p {
        outs "Hello, ";
        a(href => "/users/gugod") { "gugod" };
    }

=head1 DESCRIPTION

This is a new try to use L<Devel::Declare> to change the Perl5
language. It learns pretty much everything from L<Template::Declare>,
and has similar interface. With only one difference: how element
attributes are defined.

In L<Template::Declare>, it goes like:

    h1 {
        id is "title";
        outs "Hi";
    };

In here, it is:

    h1(id => "title") { "Hi" };

Or a shorthand for "id" attribute:

    h1("#title") { "Hi" };

There are several HTML tags named the same as one of perl builtin or ops.
Here's a list of them:

    link map q s sub tr

To generate markup with these tags, put C<"html_"> prefix in front of them,
like:

    html_q { "I a quotation, but this tag is not supported by IE." }

It'll produce:

    <q>I a quotation, but this tag is not supported by IE.</q>

For tables, since there's absolutely no way to clobber "tr" in Perl
AFAIK (not without using source filter,) tr and td are both renamed
to row and cell, correspondly:

    table {
      row {
        cell { "gugod" };
        cell { "170cm" };
        cell { "100kg" };
      }
    }

It actually makes more sense. This idea is borrowed from
L<Template::Declare>

Several helper methods are defined in L<Markapl::Helpers>. Read the
documentation there too.

=head1 INTERFACE 

=over

=item template($name, $code);

Defines a template.

=item render($name)

You need to call it as class method like,

   MyView->render("/foo.html");

Or

   render MyView, "/foo.html";

If you happen to like this style.

Doesn't support template variables yet. Stay tuned.

=item outs($str);

Should only be used inside a template body. It appends C<$str> to
current output buffer frame.

=item set($name, $value)

Store a value under given name in your view package stash. Should be used
as a class method like:

    MyView->set(title => "Greeting");

Think of it is assigning values to global template variables. C<$name>
must be a string and C<$value> can be any scalar. It could be a
reference to other structured data.

=item get($name)

Retriving a named value from the view package stash. This should only
be used in template as a function call, like:

    template 'index.html' => sub {
        h1 { get("title") }
    }

If you call it as a class method, it will still work:

    MyView->get("title")

=back

=head1 DEPENDENCIES

C<Devel::Declare>

=head1 BUGS AND LIMITATIONS

No bugs have been reported.

Please report any bugs or feature requests to
C<bug-markapl@rt.cpan.org>, or through the web interface at
L<http://rt.cpan.org>.

=head1 AUTHOR

Kang-min Liu  C<< <gugod@gugod.org> >>

=head1 LICENCE AND COPYRIGHT

Copyright (c) 2008, Kang-min Liu C<< <gugod@gugod.org> >>.

This is free software, licensed under:

    The MIT (X11) License

=head1 DISCLAIMER OF WARRANTY

BECAUSE THIS SOFTWARE IS LICENSED FREE OF CHARGE, THERE IS NO WARRANTY
FOR THE SOFTWARE, TO THE EXTENT PERMITTED BY APPLICABLE LAW. EXCEPT WHEN
OTHERWISE STATED IN WRITING THE COPYRIGHT HOLDERS AND/OR OTHER PARTIES
PROVIDE THE SOFTWARE "AS IS" WITHOUT WARRANTY OF ANY KIND, EITHER
EXPRESSED OR IMPLIED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE. THE
ENTIRE RISK AS TO THE QUALITY AND PERFORMANCE OF THE SOFTWARE IS WITH
YOU. SHOULD THE SOFTWARE PROVE DEFECTIVE, YOU ASSUME THE COST OF ALL
NECESSARY SERVICING, REPAIR, OR CORRECTION.

IN NO EVENT UNLESS REQUIRED BY APPLICABLE LAW OR AGREED TO IN WRITING
WILL ANY COPYRIGHT HOLDER, OR ANY OTHER PARTY WHO MAY MODIFY AND/OR
REDISTRIBUTE THE SOFTWARE AS PERMITTED BY THE ABOVE LICENCE, BE
LIABLE TO YOU FOR DAMAGES, INCLUDING ANY GENERAL, SPECIAL, INCIDENTAL,
OR CONSEQUENTIAL DAMAGES ARISING OUT OF THE USE OR INABILITY TO USE
THE SOFTWARE (INCLUDING BUT NOT LIMITED TO LOSS OF DATA OR DATA BEING
RENDERED INACCURATE OR LOSSES SUSTAINED BY YOU OR THIRD PARTIES OR A
FAILURE OF THE SOFTWARE TO OPERATE WITH ANY OTHER SOFTWARE), EVEN IF
SUCH HOLDER OR OTHER PARTY HAS BEEN ADVISED OF THE POSSIBILITY OF
SUCH DAMAGES.
